test/storage/index_builder_test.cpp

```diff
@@ -0,0 +1,179 @@
+  }
+
+  {
+    auto col = table->GetSqlTable()->GetSchema().GetColumn("address");
+    auto col_oid = catalog::indexkeycol_oid_t(!col.GetOid());
+
+    // FIXME(xueyuan): Seems the Schema::GetAttrSize for varlen attribute is the internal representation, thus
+    // incompatible with return value of the BlockLayout::AttrSize
+    key_cols_0.emplace_back(IndexKeyColumn(col_oid, col.GetType(), false, INT8_MAX & col.GetAttrSize()));
+  }
+
+  {
+    auto col = table->GetSqlTable()->GetSchema().GetColumn("sex");
+    auto col_oid = catalog::indexkeycol_oid_t(!col.GetOid());
+    key_cols_0.emplace_back(IndexKeyColumn(col_oid, col.GetType(), false));
+  }
+
+  auto index_key_schema = IndexKeySchema(key_cols_0);
+
+  // Build the index
+  index_factory.SetOid(index_oid_0);
+  index_factory.SetConstraintType(ConstraintType::DEFAULT);
+  index_factory.SetKeySchema(index_key_schema);
+  auto index_0 = index_factory.Build();
+
+  // Populate the index
+  bool success = index_manager_->PopulateIndex(txn_, *table->GetSqlTable(), index_0, false);
+  EXPECT_EQ(success, true);
+
+  // Create the projected row for index
+  const IndexMetadata &metadata_0 = index_0->GetIndexMetadata();
+  const auto &pr_initializer = metadata_0.GetProjectedRowInitializer();
+  auto *key_buf_index = common::AllocationUtil::AllocateAligned(pr_initializer.ProjectedRowSize());
+  ProjectedRow *index_key = pr_initializer.InitializeRow(key_buf_index);
+
+  // Create the projected row for the sql table
+  std::vector<catalog::col_oid_t> col_oids;
+  col_oids.reserve(index_key_schema.size());
+  for (const auto &it : index_key_schema) {
+    col_oids.emplace_back(catalog::col_oid_t(!it.GetOid()));
+  }
+  auto init_and_map = table->GetSqlTable()->InitializerForProjectedRow(col_oids);
+  auto *key_buf = common::AllocationUtil::AllocateAligned(init_and_map.first.ProjectedRowSize());
+  ProjectedRow *key = init_and_map.first.InitializeRow(key_buf);
+
+  // Record the col_id of each column
+  std::vector<col_id_t> sql_table_cols;
+  sql_table_cols.reserve(key->NumColumns());
+  for (uint16_t i = 0; i < key->NumColumns(); ++i) {
+    sql_table_cols.emplace_back(key->ColumnIds()[i]);
+  }
+  for (const auto &it : *table->GetSqlTable()) {
+    if (table->GetSqlTable()->Select(txn_, it, key)) {
+      for (uint16_t i = 0; i < key->NumColumns(); ++i) {
+        key->ColumnIds()[i] = index_key->ColumnIds()[i];
+      }
+      std::vector<TupleSlot> tup_slots;
+      index_0->ScanKey(*txn_, *key, &tup_slots);
+      bool flag = false;
+      for (const auto &tup_slot : tup_slots) {
+        if (it == tup_slot) {
+          flag = true;
+        }
+      }
+      EXPECT_EQ(flag, true);
+      for (uint16_t i = 0; i < key->NumColumns(); ++i) {
+        key->ColumnIds()[i] = sql_table_cols[i];
+      }
+    }
+  }
+  delete[] key_buf_index;
+  delete[] key_buf;
+  delete index_0;
+  delete table;
+}
+}  // namespace terrier::storage::index
```

Comment:
You mentioned an edge case in your design document where a key could be inserted twice into the index. Therefore, there is supposed to be a test case on that in your unit test.